test/Autofac.Test/Core/DisposerTests.cs

```diff
@@ -38,5 +39,135 @@ public void OnDispose_DisposerDisposesContainedInstances()
             disposer.Dispose();
             Assert.True(instance.IsDisposed);
         }
+
+        [Fact]
+        public void CannotAddObjectsToDisposerAfterSyncDispose()
+        {
+            var instance = new DisposeTracker();
+
+            var disposer = new Disposer();
+            disposer.AddInstanceForDisposal(instance);
+            Assert.False(instance.IsDisposed);
+            Assert.False(instance.IsDisposed);
+            disposer.Dispose();
+            Assert.True(instance.IsDisposed);
+
+            Assert.Throws<ObjectDisposedException>(() =>
+            {
+                disposer.AddInstanceForDisposal(instance);
+            });
+        }
+
+        [Fact]
+        public void DisposerDisposesOfObjectsAsyncIfIAsyncDisposableDeclared()
+        {
+            var instance = new AsyncDisposeTracker();
+
+            var disposer = new Disposer();
+            disposer.AddInstanceForDisposal(instance);
+            Assert.False(instance.IsSyncDisposed);
+            Assert.False(instance.IsAsyncDisposed);
+            var result = disposer.DisposeAsync();
```

Comment:
You can use `await` in all the tests in this fixture that call `DisposeAsync`.

Reply:
Updated; using await now instead of GetAwaiter().GetResult(). Not entirely sure why I didn't do that in the first place...